src/tests/end2end/ComputeCopyStorageBufferTests.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright 2018 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tests/NXTTest.h"
+
+#include "utils/NXTHelpers.h"
+#include <array>
+
+class ComputeCopyStorageBufferTests : public NXTTest {
+  public:
+    static constexpr int INSTANCES = 4;
+    static constexpr int UINTS_PER_INSTANCE = 4;
+    static constexpr int NUM_UINTS = INSTANCES * UINTS_PER_INSTANCE;
+
+    void BasicTest(const char* shader);
+};
+
+void ComputeCopyStorageBufferTests::BasicTest(const char* shader) {
+    auto bgl = utils::MakeBindGroupLayout(
+        device, {
+                    {0, nxt::ShaderStageBit::Compute, nxt::BindingType::StorageBuffer},
+                    {1, nxt::ShaderStageBit::Compute, nxt::BindingType::StorageBuffer},
+                });
+
+    // Set up shader and pipeline
+    auto module = utils::CreateShaderModule(device, nxt::ShaderStage::Compute, shader);
+    auto pl = utils::MakeBasicPipelineLayout(device, &bgl);
+    auto pipeline = device.CreateComputePipelineBuilder()
+                        .SetLayout(pl)
+                        .SetStage(nxt::ShaderStage::Compute, module, "main")
+                        .GetResult();
+
+    // Set up src storage buffer
+    auto src =
+        device.CreateBufferBuilder()
+            .SetSize(NUM_UINTS * sizeof(uint32_t))
+            .SetAllowedUsage(nxt::BufferUsageBit::Storage | nxt::BufferUsageBit::TransferSrc |
+                             nxt::BufferUsageBit::TransferDst)
+            .SetInitialUsage(nxt::BufferUsageBit::TransferDst)
+            .GetResult();
+    std::array<uint32_t, NUM_UINTS> expected;
+    for (uint32_t i = 0; i < NUM_UINTS; ++i) {
+        expected[i] = (i + 1u) * 0x11111111u;
+    }
+    src.SetSubData(0, sizeof(expected), reinterpret_cast<const uint8_t*>(expected.data()));
+    src.TransitionUsage(nxt::BufferUsageBit::Storage);
```

Comment:
nit: this transition is unnecessary and overriden immediately by EXPECT_BUFFER_U32_RANGE_EQ